lib/archive/diff.go

```diff
@@ -103,16 +113,28 @@ func Tar(op trace.Operation, dir string, changes []docker.Change, spec *FilterSp
 			_ = tw.WriteHeader(hdr)
 
 			p := filepath.Join(dir, change.Path)
-			if hdr.Typeflag == tar.TypeReg && hdr.Size != 0 {
+			if (hdr.Typeflag == tar.TypeReg || hdr.Typeflag == tar.TypeRegA) && hdr.Size != 0 {
 				f, err = os.Open(p)
 				if err != nil {
 					if os.IsPermission(err) {
 						err = nil
 					}
 					return
 				}
+
 				if f != nil {
+					// make sure we get out of io.Copy if context is canceled
+					done := make(chan struct{})
+					go func() {
+						select {
+						case <-op.Done():
+							f.Close()
+						case <-done:
+						}
+					}()
+
 					_, err = io.Copy(tw, f)
+					close(done)
 					if err != nil {
 						op.Errorf("Error writing archive data: %s", err.Error())
```

Comment:
We do bail out - this will trip us into the next iteration of the loop which checks context.Err()
We do end up writing the archive terminator, which maybe makes it unclear that it's a truncated archive - updated to suppress tar closure.